File: src/main/java/io/github/liyze09/arms/network/Packet.java
package io.github.liyze09.arms.network;

import org.jetbrains.annotations.NotNull;

import java.nio.ByteBuffer;

import static io.github.liyze09.arms.network.VarInt.readVarInt;
import static io.github.liyze09.arms.util.BytesUtil.readBytes;

public record Packet(int length, int id, byte[] data) {
    public static @NotNull Packet parsePacket(ByteBuffer input) {
        int length = readVarInt(input);
        var data = readBytes(input, length);
        var packID = readVarInt(ByteBuffer.wrap(data));
        return new Packet(length, packID, data);
    }
}

File: src/main/kotlin/io/github/liyze09/arms/network/packet/Packet.kt
package io.github.liyze09.arms.network.packet

import io.github.liyze09.arms.network.PackUtils
import io.netty.buffer.ByteBuf
import org.jetbrains.annotations.Contract

@JvmRecord
data class Packet(@JvmField val length: Int, @JvmField val id: Int, @JvmField val data: ByteBuf) {
    companion object {
        @Contract("_, _ -> new")
        fun of(id: Int, data: ByteBuf): Packet {
            return Packet(data.readableBytes() + PackUtils.getVarIntLength(id), id, data)
        }
    }
}
